File: src/lib/events/content-media.ts
import type { NostrEvent } from '../protocol';

export type ContentAttachment = {
  readonly url: string;
  readonly type: 'image' | 'video' | 'audio' | 'link';
  readonly aspectRatio?: string;
};

const urlPattern = /\bhttps:\/\/[^\s<>"']+/gi;

export function contentAttachments(event: NostrEvent): ContentAttachment[] {
  return dedupe([
    ...imetaAttachments(event),
    ...contentUrls(event.content).map((url) => classifyUrl(url)),
  ]);
}

export function contentUrls(content: string): string[] {
  return [...content.matchAll(urlPattern)]
    .map((match) => cleanUrl(match[0] ?? ''))
    .filter((url) => url.length > 0);
}

export function classifyUrl(url: string): ContentAttachment {
  const lower = cleanUrl(url).toLowerCase();
  if (/\.(png|jpe?g|gif|webp|avif)(\?|#|$)/.test(lower))
    return { url: cleanUrl(url), type: 'image' };
  if (/\.(mp4|webm|mov|m4v)(\?|#|$)/.test(lower))
    return { url: cleanUrl(url), type: 'video' };
  if (/\.(mp3|m4a|ogg|opus|wav|flac)(\?|#|$)/.test(lower))
    return { url: cleanUrl(url), type: 'audio' };
  return { url: cleanUrl(url), type: 'link' };
}

function imetaAttachments(event: NostrEvent): ContentAttachment[] {
  return event.tags
    .filter((tag) => tag[0] === 'imeta')
    .flatMap((tag): ContentAttachment[] => {
      const url = tagToken(tag, 'url');
      if (!url?.startsWith('https://')) return [];
      const mime = tagToken(tag, 'm')?.toLowerCase() ?? '';
      const aspectRatio = aspectRatioFromDim(tagToken(tag, 'dim'));
      if (mime.startsWith('image/'))
        return [{ url, type: 'image', aspectRatio } as const];
      if (mime.startsWith('video/'))
        return [{ url, type: 'video', aspectRatio } as const];
      if (mime.startsWith('audio/')) return [{ url, type: 'audio' }];
      return [{ ...classifyUrl(url), aspectRatio }];
    });
}

function tagToken(tag: readonly string[], name: string): string | undefined {
  const prefix = `${name} `;
  return tag.find((item) => item.startsWith(prefix))?.slice(prefix.length);
}

function cleanUrl(url: string): string {
  return url.replace(/[),.;:!?]+$/u, '');
}

function aspectRatioFromDim(dim: string | undefined): string | undefined {
  const match = dim?.match(/^(\d+)x(\d+)$/);
  if (!match) return undefined;
  const width = Number(match[1]);
  const height = Number(match[2]);
  return width > 0 && height > 0 ? `${width} / ${height}` : undefined;
}

function dedupe(items: readonly ContentAttachment[]): ContentAttachment[] {
  const seen = new Set<string>();
  return items.filter((item) => {
    if (!item.url.startsWith('https://') || seen.has(item.url)) return false;
    seen.add(item.url);
    return true;
  });
}

File: crates/lkjstr-protocol/src/content_media.rs
use crate::NostrEvent;

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum ContentAttachmentKind {
    Image,
    Video,
    Audio,
    Link,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ContentAttachment {
    pub url: String,
    pub kind: ContentAttachmentKind,
    pub aspect_ratio: Option<String>,
}

#[must_use]
pub fn content_attachments(event: &NostrEvent) -> Vec<ContentAttachment> {
    dedupe(
        imeta_attachments(&event.tags)
            .into_iter()
            .chain(
                content_urls(&event.content)
                    .into_iter()
                    .map(|url| classify_url(&url)),
            )
            .collect(),
    )
}

#[must_use]
pub fn content_urls(content: &str) -> Vec<String> {
    let bytes = content.as_bytes();
    let mut urls = Vec::new();
    let mut index = 0;
    while let Some(offset) = content[index..].find("https://") {
        let start = index + offset;
        if start > 0 && is_ascii_word(bytes[start - 1]) {
            index = start + 1;
            continue;
        }
        let end = scan_url_end(bytes, start);
        if let Some(url) = clean_https_url(&content[start..end]) {
            urls.push(url);
        }
        index = end;
    }
    urls
}

#[must_use]
pub fn classify_url(url: &str) -> ContentAttachment {
    let clean = clean_url(url);
    let lower = clean.to_lowercase();
    ContentAttachment {
        kind: classify_lower_url(&lower),
        url: clean,
        aspect_ratio: None,
    }
}

#[must_use]
pub fn embedded_media_attachments(event: &NostrEvent) -> Vec<ContentAttachment> {
    content_attachments(event)
        .into_iter()
        .filter(|item| !matches!(item.kind, ContentAttachmentKind::Link))
        .collect()
}

fn imeta_attachments(tags: &[Vec<String>]) -> Vec<ContentAttachment> {
    tags.iter()
        .filter(|tag| tag.first().is_some_and(|name| name == "imeta"))
        .filter_map(|tag| imeta_attachment(tag))
        .collect()
}

fn imeta_attachment(tag: &[String]) -> Option<ContentAttachment> {
    let url = tag_token(tag, "url").and_then(clean_https_url)?;
    let mime = tag_token(tag, "m").unwrap_or_default().to_lowercase();
    let kind = if mime.starts_with("image/") {
        ContentAttachmentKind::Image
    } else if mime.starts_with("video/") {
        ContentAttachmentKind::Video
    } else if mime.starts_with("audio/") {
        ContentAttachmentKind::Audio
    } else {
        classify_lower_url(&url.to_lowercase())
    };
    Some(ContentAttachment {
        url,
        kind,
        aspect_ratio: aspect_ratio_from_dim(tag_token(tag, "dim")),
    })
}

fn tag_token<'a>(tag: &'a [String], name: &str) -> Option<&'a str> {
    let prefix = format!("{name} ");
    tag.iter()
        .find(|item| item.starts_with(&prefix))
        .map(|item| &item[prefix.len()..])
}

fn classify_lower_url(lower: &str) -> ContentAttachmentKind {
    let path = lower.split(['?', '#']).next().unwrap_or(lower);
    if ends_with_any(path, &[".png", ".jpg", ".jpeg", ".gif", ".webp", ".avif"]) {
        ContentAttachmentKind::Image
    } else if ends_with_any(path, &[".mp4", ".webm", ".mov", ".m4v"]) {
        ContentAttachmentKind::Video
    } else if ends_with_any(path, &[".mp3", ".m4a", ".ogg", ".opus", ".wav", ".flac"]) {
        ContentAttachmentKind::Audio
    } else {
        ContentAttachmentKind::Link
    }
}

fn ends_with_any(value: &str, suffixes: &[&str]) -> bool {
    suffixes.iter().any(|suffix| value.ends_with(suffix))
}

fn aspect_ratio_from_dim(dim: Option<&str>) -> Option<String> {
    let (width, height) = dim?.split_once('x')?;
    let width = width.parse::<u32>().ok()?;
    let height = height.parse::<u32>().ok()?;
    (width > 0 && height > 0).then(|| format!("{width} / {height}"))
}

fn scan_url_end(bytes: &[u8], start: usize) -> usize {
    let mut end = start;
    while end < bytes.len()
        && !matches!(
            bytes[end],
            b' ' | b'\n' | b'\t' | b'<' | b'>' | b'"' | b'\''
        )
    {
        end += 1;
    }
    end
}

fn clean_https_url(value: &str) -> Option<String> {
    let cleaned = clean_url(value);
    (cleaned.starts_with("https://") && url::Url::parse(&cleaned).ok()?.scheme() == "https")
        .then_some(cleaned)
}

fn clean_url(value: &str) -> String {
    value
        .trim_end_matches([')', ',', '.', ';', ':', '!', '?'])
        .to_owned()
}

fn dedupe(items: Vec<ContentAttachment>) -> Vec<ContentAttachment> {
    let mut urls = Vec::<String>::new();
    let mut attachments = Vec::new();
    for item in items {
        if urls.iter().any(|url| url == &item.url) {
            continue;
        }
        urls.push(item.url.clone());
        attachments.push(item);
    }
    attachments
}

fn is_ascii_word(byte: u8) -> bool {
    byte.is_ascii_alphanumeric() || byte == b'_'
}
